### PR TITLE
[WIP] Data Distribution chain

### PR DESCRIPTION
**Pull request by ironMann**
todo:
- [ ] Comply with o2 coding rules

### CODE REVIEW

**Comment by Barthelemy** [comment]
Thank you for this PR. I have given it a try. 

On CC7 `start_StfBuilderDplAdapterTest.sh` works well. On Mac, there are crashes. 
I will continue playing with it and try the DPL bridge. I'll let you know.

**Comment by Barthelemy** [comment]
The code is nicely written and commented. 

I have found one problem so far (apart from it not working on Mac) : the directory containing the headers must be named after the module (library). Otherwise the headers are not installed. As a consequence I can't use your code in the QC because I can't include any of your headers.

**Comment by Barthelemy** [comment]
When running the DPL adapter and the sink, I get the following errors  : 
```[09:48:15][ERROR] Got a message with the same header and timeslice twice!!
[09:48:15][ERROR] Unable to relay part.
```
It is because the o2DataModelAdaptor is always passed a startTime=0...

**Comment by Barthelemy** [comment]
Paging @ktf @sawenzel @sy-c @davidrohr @shahor02 as you might be interested

**Comment by ktf** [comment]
@Barthelemy was telling me that he always gets the same timeslice as input of his DPL workflow. Could you point me to where you fill the `DataProcessingHeader`?

**Comment by ironMann** [comment]
@Barthelemy @ktf 
Mac crashes:
I will need time to investigate this since it's not the setup I'm using. Could it be that the size of the  readout region it too large? This can be changed with the `--rsize` parameter of test scripts.

DPL integration:
I used `o2DataModelAdaptor()`  since o2 messages publish by the StfBuilder consists of DataHeader and a payload. Is another header needed in the stack in this case?
Currently, I'm missing a mechanism to tell the DPL that messages are belonging to a single (Sub)TimeFrame. The DPL sink, at the StfSender input, should also be able to group messages belonging to individual TimeFrames.

**Comment by ironMann** [comment]
@mkrzewic Can you please review the first two commits with changes to the data header code.

File: DataFormats/Headers/include/Headers/DataHeader.h [hunk 1]
```diff
@@ -569,10 +609,17 @@ struct DataHeader : public BaseHeader
   bool operator==(const SerializationMethod&) const; //comparison
   void print() const; ///pretty print the contents
 
+  DataIdentifier getDataIdentifier() const noexcept
+  {
+    return DataIdentifier(this->dataDescription, this->dataOrigin);
+  }
+
```
Comment (mkrzewic) [inline DataFormats/Headers/include/Headers/DataHeader.h]: why a wrapper of copy ctor?

File: DataFormats/Headers/include/Headers/DataHeader.h [hunk 2]
```diff
@@ -500,6 +500,46 @@ struct printDataOrigin {
...
+  bool operator<(const DataIdentifier& other) const
+  {
+    if (dataDescription < other.dataDescription)
+      return true;
+    if (dataOrigin < other.dataOrigin)
+      return true;
+    return false;
+  }
```
Comment (mkrzewic) [inline DataFormats/Headers/include/Headers/DataHeader.h]: this would be useful for sorting, but how do you define (and need) ordering for headers?

**Comment by ktf** [comment]
Using `o2DataModelAdaptor()` should work, however you should make sure you have the start time and interval set according what found in the RDH, I guess. I suspect what Barth is seeing is that each part is sent as a separate timeframe due to the lacking mechanism you mention.

**Comment by ktf** [comment]
Now that we have the data distribution package, is this PR relevant at all?

**Pull request closed without merging**